### PR TITLE
WASM_X64, X86: Use templates instead of strings

### LINKED ISSUE #1409: WASM: Use templates instead of strings
For speed. See: https://github.com/lcompilers/lpython/pull/1406#discussion_r1064156854

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**
closes #1409

### CODE REVIEW

**Comment by Shaikh-Ubaid** [comment]
I found three ways to accomplish #1409. 

1. Passing a lambda function
```c++
// its benefit is that the number of arguments to be passed can be controlled
// for example add(rax, rbx) takes two arguments where as mul(rbx) takes one
template<typename F>
void handleI32Opt(F && f) {
    m_a.asm_pop_r64(X64Reg::rbx);
    m_a.asm_pop_r64(X64Reg::rax);
    f();
    m_a.asm_push_r64(X64Reg::rax);
}

// can be used as follows
void visit_I32Add() {
    handleI32Opt([&](){ m_a.asm_add_r64_r64(X64Reg::rax, X64Reg::rbx);});
}
void visit_I32Mul() {
    handleI32Opt([&](){ m_a.asm_mul_r64(X64Reg::rbx);});
}
```
2. Using template
```c++
using JumpFn = void(X86Assembler::*)(const std::string&);
template<JumpFn T>
void handle_I32Compare() {
    std::string label = std::to_string(offset);
    m_a.asm_pop_r64(X64Reg::rbx);
    m_a.asm_pop_r64(X64Reg::rax);
    m_a.asm_cmp_r64_r64(X64Reg::rax, X64Reg::rbx);

    (m_a.*T)(".compare_1" + label);

    // if the `compare` condition in `true`, jump to compare_1
    // and assign `1` else assign `0`
    m_a.asm_push_imm8(0);
    m_a.asm_jmp_label(".compare.end_" + label);
    m_a.add_label(".compare_1" + label);
    m_a.asm_push_imm8(1);
    m_a.add_label(".compare.end_" + label);
}

// can be used as follows
void visit_I32Eq() { handle_I32Compare<&X86Assembler::asm_je_label>(); }
```
3. Passing by function pointer
```c++
using F64OptFn = void(X86Assembler::*)(X64FReg, X64FReg);
void handleF64Operations(F64OptFn fn) {
    X64Reg stack_top = X64Reg::rsp;
    // load second operand into floating-point register
    m_a.asm_movsd_r64_m64(X64FReg::xmm1, &stack_top, nullptr, 1, 0);
    m_a.asm_pop_r64(X64Reg::rax); // pop the argument
    // load first operand into floating-point register
    m_a.asm_movsd_r64_m64(X64FReg::xmm0, &stack_top, nullptr, 1, 0);
    // no need to pop this operand since we need space to output back result

    (m_a.*fn)(X64FReg::xmm0, X64FReg::xmm1);

    // store float result back on stack top;
    m_a.asm_movsd_m64_r64(&stack_top, nullptr, 1, 0, X64FReg::xmm0);
}

// can be used as follows
void visit_F64Add() { handleF64Operations(&X86Assembler::asm_addsd_r64_r64); }
```

File: src/libasr/codegen/wasm_to_x64.cpp [hunk 1]
```diff
@@ -344,12 +329,12 @@ class X64Visitor : public WASMDecoder<X64Visitor>,
...
-    void visit_I32GeS() { handle_I32Compare("GtE"); }
-    void visit_I32LtS() { handle_I32Compare("Lt"); }
-    void visit_I32LeS() { handle_I32Compare("LtE"); }
-    void visit_I32Ne() { handle_I32Compare("NotEq"); }
+    void visit_I32Eq() { handle_I32Compare<&X86Assembler::asm_je_label>(); }
+    void visit_I32GtS() { handle_I32Compare<&X86Assembler::asm_jg_label>(); }
+    void visit_I32GeS() { handle_I32Compare<&X86Assembler::asm_jge_label>(); }
+    void visit_I32LtS() { handle_I32Compare<&X86Assembler::asm_jl_label>(); }
+    void visit_I32LeS() { handle_I32Compare<&X86Assembler::asm_jle_label>(); }
+    void visit_I32Ne() { handle_I32Compare<&X86Assembler::asm_jne_label>(); }
```
Comment (certik) [inline src/libasr/codegen/wasm_to_x64.cpp:337]: I would use this approach, with the template.

File: src/libasr/codegen/wasm_to_x64.cpp [hunk 2]
```diff
@@ -289,52 +289,37 @@ class X64Visitor : public WASMDecoder<X64Visitor>,
...
-        m_a.asm_mul_r64(X64Reg::rbx);
-        m_a.asm_push_r64(X64Reg::rax);
+        handleI32Opt([&](){ m_a.asm_mul_r64(X64Reg::rbx);});
     }
     void visit_I32DivS() {
-        m_a.asm_pop_r64(X64Reg::rbx);
-        m_a.asm_pop_r64(X64Reg::rax);
-        m_a.asm_div_r64(X64Reg::rbx);
-        m_a.asm_push_r64(X64Reg::rax);
+        handleI32Opt([&](){ m_a.asm_div_r64(X64Reg::rbx);});
```
Comment (certik) [inline src/libasr/codegen/wasm_to_x64.cpp:310]: If the arguments are different, such as in this case, then using the lambda function seems the best.

File: src/libasr/codegen/wasm_to_x64.cpp [hunk 3]
```diff
@@ -363,7 +348,8 @@ class X64Visitor : public WASMDecoder<X64Visitor>,
         m_a.asm_movsd_m64_r64(&stack_top, nullptr, 1, 0, X64FReg::xmm0); // store float on integer stack top;
     }
 
-    void handleF64Operations(char opt) {
+    using F64OptFn = void(X86Assembler::*)(X64FReg, X64FReg);
+    void handleF64Operations(F64OptFn fn) {
```
Comment (certik) [inline src/libasr/codegen/wasm_to_x64.cpp:352]: I would use the template.

File: src/libasr/codegen/wasm_to_x64.cpp (hunk 2 above)
Comment (Shaikh-Ubaid) [reply src/libasr/codegen/wasm_to_x64.cpp:315]: Sure, got it.

**Comment by Shaikh-Ubaid** [comment]
This is ready. Please possibly review and share feedback.

**Review by certik — APPROVED** [review]

**Pull request merged**